### ISSUE TITLE
key not found: Arrow[Bool, Int32]

### ISSUE THREAD

**Issue by magnus-madsen** [flix/flix#212]
The unification based type system is making good progress, but I seem to have run into a possible backend bug (or maybe an incomplete refactor).

Specifically, the TestParser test cases `Type.Lambda01` to `Type.Lambda04` fails with:

```
key not found: Arrow[Bool, Int32]
java.util.NoSuchElementException: key not found: Arrow[Bool, Int32]
    at scala.collection.MapLike$class.default(MapLike.scala:228)
    at scala.collection.AbstractMap.default(Map.scala:59)
    at scala.collection.MapLike$class.apply(MapLike.scala:141)
    at scala.collection.AbstractMap.apply(Map.scala:59)
    at ca.uwaterloo.flix.language.phase.Codegen$Context.ca$uwaterloo$flix$language$phase$Codegen$Context$$inner$1(Codegen.scala:111)
```

Probably due to the refactoring of types.

@mhyee If you have the time, is this something you can look at? I suspect it is an easy bug with a map not being correctly updated.

**Comment by magnus-madsen** [comment]
Please tell me if you make changes, unless they are isolated to the backend.

Also, FYI: The representation of a function `Int -> Int` is now something like: `Type.Apply(Type.Arrow, List(Type.Int))`

**Comment by mhyee** [comment]
I'll take a look over the weekend.

**Comment by mhyee** [comment]
The bug has something to do with the type system refactor.

The test case `Type.Lambda01` has the type `Bool -> Int`.

The bug is being triggered in the code generator, when the `descriptor` helper function is called. The returned string should be `(Z)I`, the JVM descriptor for a method that takes an `int` and returns a `boolean`, and should be generated by [the Type.Apply case](https://github.com/flix/flix/blob/newtyper/main/src/ca/uwaterloo/flix/language/phase/Codegen.scala#L119):

```scala
tpe match {
  case Type.Apply(Type.Arrow(l), ts) => s"(${ts.take(l - 1).map(inner).mkString})${inner(ts.last)}"
  case _ => inner(tpe)
}
```

Basically, the arguments should be converted (via a call to `inner`), concatenated, and placed between parentheses. Then the return type (the type after the last arrow) should be converted (via a call to `inner`) and then appended.

I think the code snippet above is correct (though `take(l - 1)` could be simplified to `init`), but there is something wrong with `tpe`.

Going through the debugger, the value I see for `tpe` (which is the type of `Bool -> Int`) is

```
Type.Apply(Type.Arrow(1),List(Type.Apply(Type.Arrow(2),List(Type.Bool, Type.Int32))))
```

I'm not familiar with the new changes, but this doesn't look right to me. There seems to be an extra `Apply` wrapping the type.

**Comment by magnus-madsen** [comment]
What is the meaning of 
```
def f: Bool -> Int
```
anyway?

Is it:

```
def f(): Bool -> Int
```

i.e. a function that takes zero arguments and returns a function from `Bool` to `Int`, or what?

**Comment by mhyee** [comment]
We've been punting this for a while. See #130.

**Comment by magnus-madsen** [comment]
According to that thread:

```
def f(x: Int): Int = x     // f(42) returns 42
def f: Int -> Int = x -> x // f() returns a lambda, so f()(42) returns 42
```

so 

```
def f: Bool -> Int = ??? // f() returns a function from Bool to Int.
```

The question is then how this should be represented in types. Right now, the type of a function is: Type.Apply(Type.Arrow, Type...) where Type... is a list of the arguments and return type.

So then I think `Type.Apply(Type.Arrow(1),List(Type.Apply(Type.Arrow(2),List(Type.Bool, Type.Int32))))` is correct. The part `Type.Apply(Type.Arrow(2),List(Type.Bool, Type.Int32))` is the `Bool -> Int`, and the outer part: `Type.Apply(Type.Arrow(1),List(inner))` is a function which takes zero arguments and returns inner.

In other words, I am claiming the type is correct, but then what?

**Comment by mhyee** [comment]
OK, then the question is how to represent the type on the JVM. `f` needs to be compiled as JVM method that takes no arguments and returns a functional interface (that takes a bool and returns an int). This goes back to the original problem, where a key isn't found.

The map that's being searched is the `interfaces` map (of functional interfaces), on [this line](https://github.com/flix/flix/blob/newtyper/main/src/ca/uwaterloo/flix/language/phase/Codegen.scala#L111). So it looks like a functional interface isn't being generated.

I think the fix will involve [`generateInterfaceNames`](https://github.com/flix/flix/blob/newtyper/main/src/ca/uwaterloo/flix/language/phase/LoadBytecode.scala#L168), but there's probably other little details to consider. Right now, `generateInterfaceNames` traverses the expression AST, but it's missing the case where a function/constant definition returns a function. It probably needs to traverse the types as well.

I'm not sure how much time I'll have this week, so you could give it a try or wait until the weekend.

**Comment by magnus-madsen** [comment]
OK; so this is weird. 

If I write:

```
def f: Bool -> Int = x -> 42
```

then it compiles fine.

But if I write

```
def f: Bool -> Int = ???
```
then I get the error above. Any ideas why?

**Comment by mhyee** [comment]
I'm assuming the front-end now supports lambda expressions (or at least, enough to make the first example compile).

The reason is that `generateInterfaceNames` traverses the expression AST, so when it sees `x -> 42`, it knows it has to generate a functional interface for the `Bool -> Int` type. Then when `f` is compiled, the type is correctly looked up from the map.

In the second example, there is no expression with type `Bool -> Int`, so a corresponding functional interface is never generated.

I think the fix is still the same. Because of `???`, `generateInterfaceNames` will need to traverse the types. Just traversing the expressions isn't enough.

**Comment by magnus-madsen** [comment]
OK; so we have found the bug. We need to traverse the types in the program. This doesn't seem super critical, so we can leave it open until someone gets around to it.

**Comment by magnus-madsen** [comment]
@mhyee I'm starting run into this bug again, this time related to currying. 

Is it hard to fix? Would it be possible just to generate the interfaces names on-the-fly/by-demand?

**Comment by mhyee** [comment]
This should be relatively straightforward. (But I'm not as familiar with the code now and there are also many new changes.)

One way would be to generate the interface names on demand: whenever a lambda type is encountered, generate a name for it. Then save the name, and at code generation time, generate a functional interface for each name.

Another way would be to change the `generateInterfaceNames` function to iterate over types instead of expressions. Right now, it traverses over every expression in the AST, extracts lambda types from `MkClosureRef`s, removes duplicates, and then generates a name for each type.

**Comment by magnus-madsen** [comment]
Fixed.
